src/dune_rules/compilation_context.ml

```diff
@@ -139,12 +141,24 @@ let create
   =
   let open Memo.O in
   let project = Scope.project scope in
-  let requires_compile =
+  let context = Super_context.context super_context in
+  let sandbox = Sandbox_config.no_special_requirements in
+  let* ocaml = Context.ocaml context in
+  let requires, direct =
     if Dune_project.implicit_transitive_deps project
-    then Memo.Lazy.force requires_link
-    else requires_compile
+    then Memo.Lazy.force requires_link, Resolve.Memo.return (fun _ -> true)
+    else if Version.supports_hidden_includes ocaml.version
+            && Dune_project.dune_version project >= (3, 17)
```

Comment:
I think this mode still needs to be guarded behind `implicit_transitive_deps`. Once we experiment with it more, we can enable it by default for newer compilers perhaps.

Reply:
Sorry, but I do not seem to have understood well your point. The new feature is invoked only when `implicit_transitive_deps` is false, according to the written logic, or am I missing something?

Thanks for the review.

Reply:
Yeah, that wasn't very clear at all. Let me explain the behavior I would like to achieve:

When we're using `(implicit_transitive_deps false)` and OCaml >= 5.2.0, we should use use your new implementation with the `-H` flag.

When we're using `(implicit_transitive_deps false)` and OCaml < 5.2.0, we should go back to using the old implementation of hiding `-I` flags.

When we're using `(implicit_transitive_deps true)`, it should disable using `-H` flags completely.

The issue with the current behavior is that users do expect upgrading to 3.17 to be relatively easy. With this change, they will now need to look at all their transitive deps. While I agree that it's a good thing, it's too much of a breakage to introduce in a minor version bump.

Therefore, I suggest that we guard this behind a feature flag. We already have a feature flag though `implicit_transitive_deps`, so we can just reuse it.